include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -195,24 +222,21 @@ namespace internal
     if (evaluation_flag == EvaluationFlags::nothing)
       return;
 
-    const EvaluatorVariant variant =
-      EvaluatorSelector<type, (fe_degree + n_q_points_1d > 4)>::variant;
-    using Eval = EvaluatorTensorProduct<variant,
-                                        dim,
-                                        fe_degree + 1,
-                                        n_q_points_1d,
-                                        Number>;
-    Eval eval(variant == evaluate_evenodd ?
-                shape_info.data.front().shape_values_eo :
-                shape_info.data.front().shape_values,
-              variant == evaluate_evenodd ?
-                shape_info.data.front().shape_gradients_eo :
-                shape_info.data.front().shape_gradients,
-              variant == evaluate_evenodd ?
-                shape_info.data.front().shape_hessians_eo :
-                shape_info.data.front().shape_hessians,
-              shape_info.data.front().fe_degree + 1,
-              shape_info.data.front().n_q_points_1d);
+    std::array<const MatrixFreeFunctions::UnivariateShapeData<Number> *, 3>
+      univariate_shape_data;
+
+    univariate_shape_data.fill(&shape_info.data[0]);
+
+    if (shape_info.data.size() == dim)
+      for (int i = 0; i < dim; ++i)
+        univariate_shape_data[i] = &shape_info.data[i];
+
+    Eval eval0 = create_evaluator_tensor_product(univariate_shape_data[0],
+                                                 variant == evaluate_evenodd);
+    Eval eval1 = create_evaluator_tensor_product(univariate_shape_data[1],
+                                                 variant == evaluate_evenodd);
```

Comment:
```suggestion
    Eval eval1 = create_evaluator_tensor_product(shape_info.data.size() == 1 ?
            &shape_info.data[0] : &shape_info.data[1],
                                                 variant == evaluate_evenodd);
```

Creating `univariate_shape_data` looks a bit useless. It is not used below anymore.

Reply:
```suggestion
    Eval eval2 = create_evaluator_tensor_product(shape_info.data.size() == 1 ?
        &shape_info.data[0] : (dim == 2 ? &shape_info.data[0] : &shape_info.data[2]) ,
                                                 variant == evaluate_evenodd);
```
This would require me to write something ugly like this for the third `Eval` object.

Reply:
Correct.

Reply:
For me personally how it's written currently is the more structured approach but I don't feel strongly about this at all. Let me know if i should change it.